File: src/v1smartglass/vision_one.py
"""Thin async client for the Trend Vision One v3 API.

Covers only the endpoints we need:
  - GET /v3.0/workbench/alerts          (latest workbench alerts)
  - GET /v3.0/asrm/highRiskUsers        (top risky users)
  - GET /v3.0/asrm/highRiskDevices      (top risky devices)

Docs: https://automation.trendmicro.com/xdr/api-v3
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Any

import httpx

from .config import SEVERITY_ORDER, VisionOneConfig


@dataclass(slots=True)
class Alert:
    id: str
    model: str
    severity: str
    score: int
    created_at: datetime
    impacted_entities: list[str]


@dataclass(slots=True)
class RiskEntity:
    name: str
    score: int
    kind: str  # "user" or "device"


class VisionOneClient:
    """Async client. Use as `async with VisionOneClient(cfg) as c: ...`."""

    def __init__(self, cfg: VisionOneConfig, *, client: httpx.AsyncClient | None = None):
        self._cfg = cfg
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(
            base_url=cfg.base_url,
            headers={
                "Authorization": f"Bearer {cfg.api_key}",
                "Accept": "application/json",
                "User-Agent": "v1smartglass/0.1",
            },
            timeout=httpx.Timeout(30.0, connect=10.0),
        )

    async def __aenter__(self) -> "VisionOneClient":
        return self

    async def __aexit__(self, *exc: object) -> None:
        await self.aclose()

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def fetch_alerts(self) -> list[Alert]:
        start = datetime.now(timezone.utc) - timedelta(minutes=self._cfg.lookback_minutes)
        params = {
            "startDateTime": start.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "orderBy": "createdDateTime desc",
            "top": 50,
        }
        data = await self._get("/v3.0/workbench/alerts", params=params)
        min_rank = SEVERITY_ORDER[self._cfg.min_severity]
        alerts: list[Alert] = []
        for item in data.get("items", []):
            severity = str(item.get("severity", "low")).lower()
            if SEVERITY_ORDER.get(severity, 0) < min_rank:
                continue
            alerts.append(
                Alert(
                    id=str(item.get("id", "")),
                    model=str(item.get("model") or item.get("alertName") or "alert"),
                    severity=severity,
                    score=int(item.get("score") or 0),
                    created_at=_parse_dt(item.get("createdDateTime")),
                    impacted_entities=_entity_names(item.get("impactScope") or {}),
                )
            )
        return alerts

    async def fetch_top_risky_users(self, top_n: int) -> list[RiskEntity]:
        return await self._fetch_risk("/v3.0/asrm/highRiskUsers", "user", top_n)

    async def fetch_top_risky_devices(self, top_n: int) -> list[RiskEntity]:
        return await self._fetch_risk("/v3.0/asrm/highRiskDevices", "device", top_n)

    async def _fetch_risk(self, path: str, kind: str, top_n: int) -> list[RiskEntity]:
        data = await self._get(path, params={"top": max(top_n, 10), "orderBy": "riskScore desc"})
        results: list[RiskEntity] = []
        for item in data.get("items", []):
            name = item.get("name") or item.get("userPrincipalName") or item.get("endpointName") or "?"
            results.append(RiskEntity(name=str(name), score=int(item.get("riskScore") or 0), kind=kind))
        return results[:top_n]

    async def _get(self, path: str, *, params: dict[str, Any]) -> dict[str, Any]:
        resp = await self._client.get(path, params=params)
        resp.raise_for_status()
        return resp.json()


def _parse_dt(value: Any) -> datetime:
    if isinstance(value, datetime):
        return value
    if not value:
        return datetime.now(timezone.utc)
    return datetime.fromisoformat(str(value).replace("Z", "+00:00"))


def _entity_names(scope: dict[str, Any]) -> list[str]:
    names: list[str] = []
    for bucket in ("users", "accounts"):
        for entry in scope.get(bucket) or []:
            if n := entry.get("name") or entry.get("userPrincipalName"):
                names.append(str(n))
    for entry in scope.get("entities") or scope.get("desktops") or []:
        if n := entry.get("entityValue") or entry.get("name"):
            names.append(str(n))
    return names

File: even-hub/v1smartglass/src/vision-one.ts
// Vision One v3 API client for the Hub app. Goes through /__v1proxy so the
// Bearer token never lives in browser-visible request URLs and CORS is not
// an issue.

export type Region = 'us' | 'eu' | 'jp' | 'sg' | 'au' | 'in' | 'uae'

export const REGION_HOSTS: Record<Region, string> = {
  us: 'api.xdr.trendmicro.com',
  eu: 'api.eu.xdr.trendmicro.com',
  jp: 'api.xdr.trendmicro.co.jp',
  sg: 'api.sg.xdr.trendmicro.com',
  au: 'api.au.xdr.trendmicro.com',
  in: 'api.in.xdr.trendmicro.com',
  uae: 'api.uae.xdr.trendmicro.com',
}

export type Severity = 'low' | 'medium' | 'high' | 'critical'
const SEVERITY_RANK: Record<Severity, number> = { low: 0, medium: 1, high: 2, critical: 3 }

export type Alert = {
  id: string
  model: string
  severity: Severity
  score: number
  createdAt: string
  impactedEntities: string[]
}

export type RiskEntity = {
  name: string
  score: number
  kind: 'user' | 'device'
}

export type VisionOneConfig = {
  apiKey: string
  region: Region
  lookbackMinutes: number
  minSeverity: Severity
}

const PROXY_PATH = '/__v1proxy'

async function proxyGet(cfg: VisionOneConfig, path: string, params: Record<string, string | number>): Promise<unknown> {
  const target = new URL(`https://${REGION_HOSTS[cfg.region]}${path}`)
  for (const [k, v] of Object.entries(params)) {
    target.searchParams.set(k, String(v))
  }
  const resp = await fetch(`${PROXY_PATH}?url=${encodeURIComponent(target.toString())}`, {
    headers: { 'X-V1-Auth': cfg.apiKey },
  })
  if (!resp.ok) {
    const body = await resp.text()
    throw new Error(`Vision One ${resp.status}: ${body.slice(0, 200)}`)
  }
  return resp.json()
}

export async function fetchAlerts(cfg: VisionOneConfig): Promise<Alert[]> {
  const start = new Date(Date.now() - cfg.lookbackMinutes * 60_000).toISOString().replace(/\.\d+Z$/, 'Z')
  const data = (await proxyGet(cfg, '/v3.0/workbench/alerts', {
    startDateTime: start,
    orderBy: 'createdDateTime desc',
    top: 50,
  })) as { items?: unknown[] }

  const minRank = SEVERITY_RANK[cfg.minSeverity]
  const items = Array.isArray(data.items) ? data.items : []
  return items
    .map(toAlert)
    .filter((a): a is Alert => a !== null && SEVERITY_RANK[a.severity] >= minRank)
}

export async function fetchTopRiskyUsers(cfg: VisionOneConfig, topN: number): Promise<RiskEntity[]> {
  return fetchRisk(cfg, '/v3.0/asrm/highRiskUsers', 'user', topN)
}

export async function fetchTopRiskyDevices(cfg: VisionOneConfig, topN: number): Promise<RiskEntity[]> {
  return fetchRisk(cfg, '/v3.0/asrm/highRiskDevices', 'device', topN)
}

async function fetchRisk(
  cfg: VisionOneConfig,
  path: string,
  kind: 'user' | 'device',
  topN: number,
): Promise<RiskEntity[]> {
  const data = (await proxyGet(cfg, path, { top: Math.max(topN, 10), orderBy: 'riskScore desc' })) as {
    items?: unknown[]
  }
  const items = Array.isArray(data.items) ? data.items : []
  return items
    .map((raw) => toRiskEntity(raw, kind))
    .filter((r): r is RiskEntity => r !== null)
    .slice(0, topN)
}

function toAlert(raw: unknown): Alert | null {
  if (!raw || typeof raw !== 'object') return null
  const r = raw as Record<string, unknown>
  const severity = String(r.severity ?? 'low').toLowerCase() as Severity
  if (!(severity in SEVERITY_RANK)) return null

  const scope = (r.impactScope ?? {}) as Record<string, unknown>
  const entities = [
    ...((scope.users as Array<Record<string, unknown>>) ?? []).map((u) => String(u.name ?? u.userPrincipalName ?? '')),
    ...((scope.entities as Array<Record<string, unknown>>) ?? []).map((e) => String(e.entityValue ?? e.name ?? '')),
  ].filter(Boolean)

  return {
    id: String(r.id ?? ''),
    model: String(r.model ?? r.alertName ?? 'alert'),
    severity,
    score: Number(r.score ?? 0),
    createdAt: String(r.createdDateTime ?? ''),
    impactedEntities: entities,
  }
}

function toRiskEntity(raw: unknown, kind: 'user' | 'device'): RiskEntity | null {
  if (!raw || typeof raw !== 'object') return null
  const r = raw as Record<string, unknown>
  const name = r.name ?? r.userPrincipalName ?? r.endpointName
  if (!name) return null
  return { name: String(name), score: Number(r.riskScore ?? 0), kind }
}

export function rankAlerts(alerts: Alert[]): Alert[] {
  return [...alerts].sort((a, b) => {
    const sevDiff = SEVERITY_RANK[b.severity] - SEVERITY_RANK[a.severity]
    if (sevDiff !== 0) return sevDiff
    return (b.createdAt ?? '').localeCompare(a.createdAt ?? '')
  })
}
